cmd/run.go

```diff
@@ -43,7 +44,15 @@ to the function's source.  Use --build to override this behavior.
 
 `,
 		SuggestFor: []string{"rnu"},
-		PreRunE:    bindEnv("build", "path", "registry"),
+		PreRunE:    bindEnv("build", "path", "registry", "verbose"),
```

Comment:
What about making the help text on lines 20-21 use `container` similarly to the others. E.g. `run the function container locally`?

Reply:
Yes, that's a good question:  because from the perspective of a Function developer, they are just running the function.  Additionally, we may want to consider updating the `run` command to support both running on the host or containerized.  So I am trying to be more deliberate when using the word "container" to only be for situations where it is explicitly exposing this concept to the user:  Build.  Of course, that's just my rationale, which indeed might be misguided.

Reply:
> Additionally, we may want to consider updating the `run` command to support both running on the host or containerized.

I like this idea. I wonder how it would be best manifested. Each runtime has their own way of starting the function locally - and some don't yet explicitly support this. Of course the buildpacks know these things e.g. `npm start` is common for Node.js - so we haven't really considered a common runtime "start" command that does things "natively". My first instinct is to say that it could be indicated in manifest.yaml with sane defaults for already-known/supported runtimes. But I'm aware there exist ways to deal with this question already, e.g. Procfiles. See for example https://github.com/knative/func/blob/main/templates/python/http/Procfile. Something like `func fun --native` would be really nice if we can work out how to best indicate that in a project's metadata - and again with sane defaults for built-in runtimes.

Reply:
Yes, this sounds a lot like what I was thinking of as well @lance.  For example:

Consider how templates work during the initialization phase, now imagine something almost exactly the same but during the build phase; let's call them 'scaffolding'.  Embedded into the binary are the scaffolding for language runtimes we natively support, and they could be included in the language repositories for extensibility.
When a Function is built (either from `build` or `deploy`), the contents of the scaffolding is written into `.func/runs/0/` (or we could use a temp directory, but I think developers might appreciate having it close at hand for inspection).  This scaffolding code consists of everything necessary to run their function _except_ the function code itself.  In the case of Go, this would be `main.go`, which would then import the user's Function code in the relative directory `./function`.  This would be (from the project root) `myFunction/.func/runs/0/function` which itself would be a simple symlink `-> ../../../`.  Included in the scaffolding would be the command, perhaps in the form of a Procfile, which would be able to actually invoke the code.  Upon completion of the run (when the process is canceled), the run directory `.func/runs/0` is removed unless the user specifically requests it be retained using, for example the `-k|--keep` flag.

This allows our Functions to remain dependency-free (and thus gain all the future benefits of a proper FaaS, such as auto-upgrades, etc) without losing the ability to run locally.  It also keeps things quite simple for developers who want to stay at a high level, working mostly in their libraries/code.  For those that want to take lower-level control, or to enable CI/CD to run these "pure" dependency-free functions, this process can be exposed as `func scaffold`:

Should a user decide they would rather not have our system's scaffolding, or take control of it themselves, they could run `func scaffold`, which would write the scaffolding code into the project source-controlled tree (perhaps `./run/...`) and a flag added to func.yaml such as `scaffold=false` indicating the Function developer has taken control of the code into their own hands (and will presumably check in the `/run` directory to source control) and the build/deploy commands will skip this step and instead invoke the Function using the process-wrapping code (main) in `./run`.

For environments like Pipelines, they need only `func scaffold` the Function source code once pulled from Git (no need to commit the result) to convert it into a form which it supports  (a Function wrapped in a process-creating boundary main).

Then, of course a `func run --host` would still scaffold as before, but it could skip creating the container and just straight to invoking the code in `./func/runs/0`.

This would have the added benefit of pulling all of our scaffolding into one repository where we can more easily iterate and test it as well.

Reply:
I think maybe this is a good [topic for a discussion](https://github.com/knative/func/discussions/1589)...